pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/inv/RangeIndexCreator.java

```diff
@@ -0,0 +1,471 @@
+      throws IOException {
+    _valueType = valueType;
+    String columnName = fieldSpec.getName();
+    _invertedIndexFile = new File(indexDir, columnName + BITMAP_RANGE_INDEX_FILE_EXTENSION);
+    _forwardIndexValueBufferFile = new File(indexDir, columnName + FORWARD_INDEX_VALUE_BUFFER_SUFFIX);
+    _docIdBufferFileForSorting = new File(indexDir, columnName + DOC_ID_VALUE_BUFFER_SUFFIX);
+    _numValues = fieldSpec.isSingleValueField() ? numDocs : numValues;
+    _useMMapBuffer = _numValues > NUM_VALUES_THRESHOLD_FOR_MMAP_BUFFER;
+    int valueSize = valueType.size();
+    try {
+      //use DEFAULT_NUM_RANGES if numRanges is not set
+      if(numRanges < 0) {
+        numRanges = DEFAULT_NUM_RANGES;
+      }
+      if(numDocsPerRange < 0) {
+        _numDocsPerRange = (int) Math.ceil(_numValues / numRanges);
+      }
+
+      //FORWARD INDEX - stores the actual values added
+      _forwardIndexValueBuffer = createTempBuffer((long) _numValues * valueSize, _forwardIndexValueBufferFile);
+
+      switch (_valueType) {
+        case INT:
+          _valueBuffer = new IntValueBuffer(_forwardIndexValueBuffer);
+          break;
+        case FLOAT:
+          _valueBuffer = new FloatValueBuffer(_forwardIndexValueBuffer);
+          break;
+        case LONG:
+          _valueBuffer = new LongValueBuffer(_forwardIndexValueBuffer);
+          break;
+        case DOUBLE:
+          _valueBuffer = new DoubleValueBuffer(_forwardIndexValueBuffer);
+          break;
+        default:
+          throw new UnsupportedOperationException("Range index is not supported for columns of data type:" + valueType);
+      }
+
+      //Stores the docId - this will be sorted based on the values in FORWARD INDEX in the end
+      _docIdValueBuffer = createTempBuffer((long) _numValues * Integer.BYTES, _forwardIndexValueBufferFile);
+      _docIdBuffer = new IntValueBuffer(_docIdValueBuffer);
+    } catch (Exception e) {
+      destroyBuffer(_forwardIndexValueBuffer, _forwardIndexValueBufferFile);
+      destroyBuffer(_forwardIndexValueBuffer, _docIdBufferFileForSorting);
+      throw e;
+    }
+  }
+
+  @Override
+  public void add(int dictId) {
+    _valueBuffer.put(_nextDocId, dictId);
+    _docIdBuffer.put(_nextDocId, _nextDocId);
+    _nextDocId = _nextDocId + 1;
+  }
+
+  @Override
+  public void add(int[] dictIds, int length) {
+    for (int i = 0; i < length; i++) {
+      int dictId = dictIds[i];
+      _valueBuffer.put(_nextDocId, dictId);
+      _docIdBuffer.put(_nextValueId, _nextDocId);
+      _nextValueId = _nextValueId + 1;
+    }
+  }
+
+  @Override
+  public void addDoc(Object document, int docIdCounter) {
+    throw new IllegalStateException("Bitmap inverted index creator does not support Object type currently");
+  }
+
+  @Override
+  public void seal()
+      throws IOException {
+    //sort the forward index copy
+    //go over the sorted value to compute ranges
+    IntComparator comparator = (i, j) -> {
```

Comment:
So this is either going to sort raw values or dictIds. Correct?

Reply:
The actual sort happens on docIdBuffer but uses forward index for sorting

Reply:
Let's not call it forward index. Let's call it keys?